test/extensions/dynamic_modules/http/abi_impl_test.cc

```diff
@@ -9,267 +9,389 @@ namespace Extensions {
+                                                 testing::Eq(0), testing::Eq("dynamic_module")));
+  envoy_dynamic_module_callback_http_send_response(filter_.get(), 200, nullptr, 3, nullptr, 0);
+}
+
+TEST_F(DynamicModuleHttpFilterTest, SendResponseEmptyResponse) {
+  Http::TestResponseHeaderMapImpl response_headers;
+  filter_->response_headers_ = &response_headers;
+
+  // Test with empty response.
+  EXPECT_CALL(decoder_callbacks_, sendLocalReply(Envoy::Http::Code::OK, testing::Eq(""), _,
+                                                 testing::Eq(0), testing::Eq("dynamic_module")));
+  EXPECT_CALL(decoder_callbacks_, encodeHeaders_(_, _));
+
+  envoy_dynamic_module_callback_http_send_response(filter_.get(), 200, nullptr, 3, nullptr, 0);
+}
+
+TEST_F(DynamicModuleHttpFilterTest, SendResponse) {
+  std::list<std::pair<std::string, std::string>> headers = {
+      {"single", "value"}, {"multi", "value1"}, {"multi", "value2"}};
+  size_t header_count = headers.size();
+  envoy_dynamic_module_type_module_http_header* header_array =
+      new envoy_dynamic_module_type_module_http_header[header_count];
+
+  size_t index = 0;
+  for (const auto& [key, value] : headers) {
+    header_array[index].key_length = key.size();
+    header_array[index].key_ptr = const_cast<char*>(key.c_str());
+    header_array[index].value_length = value.size();
+    header_array[index].value_ptr = const_cast<char*>(value.c_str());
+    ++index;
   }
+  EXPECT_CALL(decoder_callbacks_, sendLocalReply(Envoy::Http::Code::OK, testing::Eq(""), _,
+                                                 testing::Eq(0), testing::Eq("dynamic_module")));
+  EXPECT_CALL(decoder_callbacks_, encodeHeaders_(_, _)).WillOnce(Invoke([](auto& headers, auto) {
+    EXPECT_EQ(headers.get(Http::LowerCaseString("single"))[0]->value().getStringView(), "value");
+    EXPECT_EQ(headers.get(Http::LowerCaseString("multi"))[0]->value().getStringView(), "value1");
+    EXPECT_EQ(headers.get(Http::LowerCaseString("multi"))[1]->value().getStringView(), "value2");
+  }));
+
+  envoy_dynamic_module_callback_http_send_response(filter_.get(), 200, header_array, header_count,
+                                                   nullptr, 0);
+}
+
+TEST_F(DynamicModuleHttpFilterTest, SendResponseWithBody) {
+  std::list<std::pair<std::string, std::string>> headers = {
+      {"single", "value"}, {"multi", "value1"}, {"multi", "value2"}};
+  size_t header_count = headers.size();
+  envoy_dynamic_module_type_module_http_header* header_array =
+      new envoy_dynamic_module_type_module_http_header[header_count];
+
+  size_t index = 0;
+  for (const auto& [key, value] : headers) {
+    header_array[index].key_length = key.size();
+    header_array[index].key_ptr = const_cast<char*>(key.c_str());
+    header_array[index].value_length = value.size();
+    header_array[index].value_ptr = const_cast<char*>(value.c_str());
+    ++index;
+  }
+
+  const std::string body_str = "body";
```

Comment:
could you add the test case where body is empty (like nullptr)?

Reply:
that is what the test above it is. unless you're asking for specifically an empty string test.

Reply:
aha ok now i found it - sorry about that!